src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -140,38 +140,64 @@ internal void Freeze()
             _orderedSynthesizedMethods = _synthesizedMethods.OrderBy(kvp => kvp.Key).Select(kvp => kvp.Value).AsImmutable();
 
             // Sort proxy types.
-            _orderedProxyTypes = _proxyTypes.OrderBy(kvp => kvp.Key).Select(kvp => kvp.Value).AsImmutable();
+            _orderedProxyTypes = _proxyTypes.OrderBy(kvp => kvp.Key.Size).Select(kvp => kvp.Value).AsImmutable();
         }
 
         private bool IsFrozen => _frozen != 0;
 
-        internal Cci.IFieldReference CreateDataField(ImmutableArray<byte> data)
+        /// <summary>
+        /// Gets a field that can be used to cache an array allocated to store data from a corresponding <see cref="CreateDataField"/> call.
+        /// </summary>
+        /// <param name="data">The data that will be used to initialize the field.</param>
+        /// <param name="arrayType">The type of the field, e.g. int[].</param>
+        /// <returns>The field to use to cache an array for this data and alignment.</returns>
+        internal Cci.IFieldReference CreateArrayCachingField(ImmutableArray<byte> data, Cci.IArrayTypeReference arrayType, SpecialType elementType)
         {
             Debug.Assert(!IsFrozen);
-            Cci.ITypeReference type = _proxyTypes.GetOrAdd((uint)data.Length, GetStorageStruct);
-            return _mappedFields.GetOrAdd(data, data0 =>
-            {
-                var name = GenerateDataFieldName(data0);
-                var newField = new MappedField(name, this, type, data0);
-                return newField;
-            });
-        }
 
-        private Cci.ITypeReference GetStorageStruct(uint size)
+            // Create a dedicated mapped field for the array type, separate from the data that'll be stored into that array.
+            // Call sites will lazily instantiate the array to cache in this field, rather than forcibly instantiating
+            // all of them when the private implementation details class is first used.
+            return _mappedFields.GetOrAdd((data, 1, elementType), key =>
+                new MappedField(GenerateDataFieldName(key.Data, key.Alignment, key.ElementType), this, arrayType, block: default));
```

Comment:
>MappedField [](http://example.com/codeflow?start=20&length=11)

It is quite possible that I am missing something, but it feels inappropriate to to use this type for a field that is not associated with a blob.

Reply:
I can change the comment on the type (e.g. "Definition of a simple field mapped to a metadata block or that stores an array used to cache such a metadata block") or change the name of the type.  But it seemed like unnecessary complication to duplicate the type and the collection and and their handling; this is just expanding the use of something that was originally there for a single purpose and now that purpose is expanded.  If you really believe it'd be better to duplicate it all, though, it's your code base, I of course can.

Reply:
I don't think we should be reusing that type. Both from the clarity and instance size perspective.

Reply:
I don't agree, but it's your code base, so I won't argue.